### PR TITLE
PyrakoonClient can use multiple clients to improve throughput in threaded situations (like the DAL)

### PR DESCRIPTION
**Pull request by JeffreyDevloo**

### CODE REVIEW

File: packaging/debian/debian/control [hunk 1]
```diff
@@ -7,6 +7,7 @@ Build-Depends: python (>= 2.7.2), debhelper (>= 9)
 
 Package: openvstorage-extensions
 Architecture: amd64
-Pre-Depends: python (>= 2.7.2), python-paramiko, python-rpyc, python-protobuf, python-requests (>=2.9.1), python-ujson
+Pre-Depends: python (>= 2.7.2), python-gevent (>= 0.13.0-1build2), python-paramiko, python-rpyc, python-protobuf,
```
Comment (toolslive) [inline packaging/debian/debian/control]: no `(< 3.0.0)` ?

File: src/db/arakoon/pyrakoon/client/base_client.py [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+        :param keys: All keys to fetch
+        :type keys" List[str]
+        :param must_exist: Should all listed keys exist
+        :type must_exist: bool
+        :return: Generator that yields key value pairs
+        :rtype: iterable[Tuple[str, any]
+        """
+        raise NotImplementedError()
+
+    def set(self, key, value, transaction=None):
```
Comment (toolslive) [inline src/db/arakoon/pyrakoon/client/base_client.py]: as stated in earlier discussions, I'm not a fan of the `transaction = None`) parameter, especially given that the parameter added here is a string. You might as well have the sequence as a parameter, or use the comfort calls on the sequence and remove this parameter altogether. With a sequence, it's absolutely clear no communication with the server is happening at all, and what the allowed set of scenarios is. Now it's less so.

File: src/db/arakoon/pyrakoon/client/base_client.py [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+    def _next_key(key):
+        # type: (str) -> str
+        """
+        Calculates the next key (to be used in range queries)
+        :param key: Key to calucate of
+        :type key: str
+        :return: The next key
+        :rtype: str
+        """
+        encoding = 'ascii'  # For future python 3 compatibility
```
Comment (toolslive) [inline src/db/arakoon/pyrakoon/client/base_client.py]: isn't that an illusion?

File: src/db/arakoon/pyrakoon/client/base_client.py [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+        array = bytearray(str(key), encoding)
+        for index in range(len(array) - 1, -1, -1):
+            array[index] += 1
+            if array[index] < 128:
+                while array[-1] == 0:
+                    array = array[:-1]
+                return str(array.decode(encoding))
+            array[index] = 0
+        return '\xff'
```
Comment (toolslive) [inline src/db/arakoon/pyrakoon/client/base_client.py]: is this correct?

File: src/db/arakoon/pyrakoon/client/base_client.py [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+        :rtype: PyrakoonLock
+        """
+        raise NotImplementedError()
+
+    def apply_callback_transaction(self, transaction_callback, max_retries=0, retry_wait_function=None):
+        # type: (callable, int, callable) -> None
+        """
+        Apply a transaction which is the result of the callback.
+        The callback should build the complete transaction again to handle the asserts. If the possible previous run was interrupted,
+        the Arakoon might only have partially applied all actions therefore all asserts must be re-evaluated
```
Comment (toolslive) [inline src/db/arakoon/pyrakoon/client/base_client.py]: partially?

File: src/db/arakoon/pyrakoon/client/client_pool.py [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+class PyrakoonPool(object):
+    """
+    Pyrakoon pool.
+    Keeps a number of Pyrakoon clients queued up to avoid waiting too long on a socket lock of a single instance
+    Uses PyrakoonClient as it has retries on master loss
+    """
+
+    _logger = Logger('extensions')
+
+    # Frequency at which the pool is populated at startup
```
Comment (toolslive) [inline src/db/arakoon/pyrakoon/client/client_pool.py]: `at -> with`

File: src/db/arakoon/pyrakoon/client/client_pool.py [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+        """
+        return PyrakoonClient(*self._pyrakoon_args)
+
+    def _add_client(self):
+        # type: () -> None
+        """
+        Add a new client to the pool
+        :return: None
+        """
+        sleep_time = 0.1
```
Comment (toolslive) [inline src/db/arakoon/pyrakoon/client/client_pool.py]: what does this have to do with the `SPAWN_FREQUENCY` ?

File: src/db/arakoon/pyrakoon/client/client_pooled.py [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+            time.sleep(random.randint(0, 25) / 100.0)
+
+        retry_wait_func = retry_wait_function or default_retry_wait
+        tries = 0
+        success = False
+        while success is False:
+            tries += 1
+            try:
+                transaction = transaction_callback()  # type: str
+                return self.apply_transaction(transaction)
```
Comment (toolslive) [inline src/db/arakoon/pyrakoon/client/client_pooled.py]: why have a `succes` variable if you never set it to `True` ?

File: src/db/arakoon/pyrakoon/client/base_client.py (hunk 2 above)
Comment (JeffreyDevloo) [reply src/db/arakoon/pyrakoon/client/base_client.py]: Out of scope for the current changes really. It will require some refactor work

File: src/db/arakoon/pyrakoon/client/base_client.py (hunk 5 above)
Comment (JeffreyDevloo) [reply src/db/arakoon/pyrakoon/client/base_client.py]: As in get interrupted mid sequence because of an assert and not do anything at all.
It's bad phrasing

File: src/db/arakoon/pyrakoon/client/client_pool.py (hunk 7 above)
Comment (JeffreyDevloo) [reply src/db/arakoon/pyrakoon/client/client_pool.py]: Its just a retry-interval
